engine/buildkit/client.go

```diff
@@ -452,6 +492,121 @@ func withDescHandlerCacheOpts(ctx context.Context, ref bkcache.ImmutableRef) con
 	})
 }
 
+func (c *Client) ListenHostToContainer(
+	ctx context.Context,
+	hostListenAddr, proto, upstream string,
+) (*session.ListenResponse, func() error, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get requester session ID: %s", err)
+	}
+
+	clientCaller, err := c.SessionManager.Get(ctx, clientMetadata.ClientID, false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get requester session: %s", err)
+	}
+
+	conn := clientCaller.Conn()
+
+	tunnelClient := session.NewTunnelListenerClient(conn)
+
+	listener, err := tunnelClient.Listen(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to listen: %s", err)
+	}
+
+	err = listener.Send(&session.ListenRequest{
+		Addr:     hostListenAddr,
+		Protocol: proto,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to send listen request: %s", err)
+	}
+
+	listenRes, err := listener.Recv()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to receive listen response: %s", err)
+	}
+
+	conns := map[string]net.Conn{}
+	connsL := &sync.Mutex{}
+	sendL := &sync.Mutex{}
+
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			res, err := listener.Recv()
+			if err != nil {
+				bklog.G(ctx).Warnf("listener recv err: %s", err)
+				return
+			}
+
+			connID := res.GetConnId()
+			if connID == "" {
+				continue
+			}
+
+			connsL.Lock()
+			conn, found := conns[connID]
+			connsL.Unlock()
+
+			if !found {
+				conn, err := c.dialer.Dial(proto, upstream)
+				if err != nil {
+					bklog.G(ctx).Warnf("failed to dial %s %s: %s", proto, upstream, err)
+					return
+				}
+
+				connsL.Lock()
+				conns[connID] = conn
+				connsL.Unlock()
+
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+
+					for {
+						data := make([]byte, 1024) // TODO larger?
```

Comment:
nit: I _think_ we can safely re-use the buffer since we always truncate it to `[:n]`. Not that we need to optimize performance right now, but if it's as easy as just moving `data := ...` above the `for {` then may as well since networking stuff is obviously a bit more performance sensitive than the rest of our codebase.

If it's not as easy in practice as I think, then no worries.